README
NAME
    forkprove - forking prove

SYNOPSIS
      forkprove -MMoose -MCatalyst -MDBIx::Class -j8 -lr t

DESCRIPTION
    forkprove is a forking version of prove that allows you to preload
    modules with "-M" and then run each test after forking, potentially
    makes most of the tests to run faster without an overhead of loading the
    same modules again and again, while having an isolated, clean testing
    environment with a fork.

    It works well with parallel testing as well, combined with "-j" option.

CONFIGURATION
    You can specify an environment variable "PERL_FORKPROVE_IGNORE" to
    contain a regular expression to ignore files to test with forkprove.

      PERL_FORKPROVE_IGNORE=no-preload forkprove -lr t

    This will run the tests whose filenames and directory match "no-preload"
    with the standard perl execution via prove.

CAVEATS
    Following kind of tests might fail when tested under forkprove. There
    are workarounds in forkprove to detect/fix some of them, so don't be
    surprised if your tests still pass :)

    *   Tests that rely on the value of $FindBin::Bin.

    *   Tests that relies on "CHECK" and "INIT" block to run correctly.

    See also
    <http://search.cpan.org/~flora/Test-Aggregate-0.364/lib/Test/Aggregate.p
    m#CAVEATS>, many of which could apply to forkprove as well.

WHY
    Test::Aggregate allows you to create a nested TAP output by running a
    whole bunch of ".t" files from a directory.

    forkprove shares the basic idea, but it creates a whole new test
    environment by making a copy of the process with fork for each test,
    which means it could run your tests that rely on the singleton state
    etc. without modifying the code to run with Test::Aggregate.

TIPS
    *   If you have a bunch of modules you want to load, put them in a file
        "t/Preload.pm" and load it:

          forkprove -Mt::Preload

        If you're lazy and just load all of ".pm" files in lib and t: use
        lib::require::all

          forkprove -Mlib::require::all=lib,t

PERFORMANCE
    Your mileage might vary, if forkprove runs your tests faster or not. For
    lightweight modules, we've found that using forkprove doesn't make any
    siginificant difference, due to the fact that fork overhead nullifies
    the benefit of preloading modules.

    On the other hand, test suite with Moose, Catalyst and Dist::Zilla seems
    to run 80-100% faster with forkprove and "-jN" option.

AUTHOR
    Tatsuhiko Miyagawa

SEE ALSO
    prove, TAP::Parser



Changes
Revision history for Perl extension App::ForkProve

{{$NEXT}}

v0.4.10  2014-06-20 11:20:07 PDT
        - Non-zero exit codes should also cause failure, even if not indicated
          by the TAP (reported by shumphrey)

v0.4.9  2013-07-03 16:04:18 PDT
        - ensure DATA is reset for preloaded .pm modules (doy)

v0.4.8  2013-06-20 22:29:08 PDT
        - Fix Test::More dependencies for subtest (doy)

v0.4.7  2013-04-07 01:59:21 PDT
        - Switched to Module::Build for now because of https://github.com/rjbs/dist-zilla/pull/164

v0.4.6  2013-04-07 01:25:09 PDT
        - Repacked with Milla and it fixes bad NAME

v0.4.5 Mon Jan  7 15:21:29 PST 2013
       - Enabled option bundling so that forprove -lr etc should work (doy)
       - Blacklisted Test::LeakTrace because it doesn't work under forkprove (doy)
       - Fixed a warning for older perls

v0.4.4 Thu Dec 20 18:22:51 JST 2012
       - Restore DATA in test script (hanabokuro)
       - Restore Getopt::Long defaults in case apps and modules in tests use it (hanabokuro)

v0.4.3 Wed Nov 21 08:11:51 PST 2012
       - Fix exit status code

v0.4.2 Mon Nov 12 20:18:49 PST 2012
       - Support --merge (#9)
       - Fix Getopt::Long options to make options case sensitive

v0.4.1 Fri Nov  9 14:10:47 PST 2012
       - Clear @ARGV for each test (#8)

v0.4.0 Fri Nov  9 08:53:58 PST 2012
       - Simplified the blacklist logic (mst)
       - Added the srand() initialization (markstos)
       - Fixed the version number scheme (xdg)

0.3.0 Thu Nov  8 16:13:17 PST 2012
      - Now you can preload Test::Builder and friends because forkprove resets the state
      - Blacklisted Test::SharedFork for now since we can't reset its state

0.2.3 Thu Nov  8 12:00:23 PST 2012
      - Added a warning when preloading modules ended up requiring Test::Builder
      - Fixed the test handler routine to use the same as prove so that .pl test can run

0.2.2 Wed Nov  7 23:31:55 PST 2012
      - Correct the error handling inside .t in case they return non-true value
      - Autoflush STDOUT/STDERR so that we can see progressive test output (doy)

0.2.1 Wed Nov  7 20:33:54 PST 2012
      - Improved error handling in .t (jswartz)
      - Relaxed the documentaion about known tests to fail

0.2.0 Wed Nov  7 12:40:48 PST 2012
      - More stablized -j option
      - Added PERL_FORKPROVE_IGNORE environment value to specify tests to ignore and
        run with the standard perl execution with prove
      - Reset regular expression engine state (doy)

0.1.5 Tue Nov  6 17:24:34 PST 2012
      - Handle -I, -l parameters in itself (doy)
      - Localize $? in running Test::Builder's END (doy)

0.1.4 Tue Nov  6 15:34:06 PST 2012
      - Probably support -j parallel testing (Thanks to doy)

0.1.3 Tue Nov  6 14:50:09 PST 2012
      - Remove unused Capture::Tiny code (doy)

0.1.2 Tue Nov  6 13:58:06 PST 2012
      - Improved the way it captures output by reopening STDOUT
      - Fixed the dependency to parent

0.1.1 Tue Nov  6 12:47:54 PST 2012
      - Fixed a bug where source handler was not correctly registered
      - Added documentation

0.1.0 Tue Nov  6 12:13:18 2012
      - original version


t/Rand.pm
package t::Rand;

srand();

1;


t/TestClass.pm
package t::TestClass;
use base qw(Test::Class);

use Test::More;
sub test_ok : Test { ok 1 }

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "forking prove",
  "dist_author" => [
    "Tatsuhiko Miyagawa <miyagawa\@bulknews.net>"
  ],
  "dist_name" => "forkprove",
  "dist_version" => "v0.4.7",
  "license" => "perl",
  "module_name" => "forkprove",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "App::Prove" => "3.25",
    "TAP::Harness" => "3.25",
    "parent" => "0.221",
    "perl" => "5.008001",
    "version" => "0.77"
  },
  "script_files" => [
    "script/forkprove"
  ],
  "test_requires" => {
    "Test::More" => "0.88",
    "Test::Requires" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/App/ForkProve.pm
package App::ForkProve;

use strict;
use 5.008_001;
use version; our $VERSION = "v0.4.9";

use App::Prove;
use Getopt::Long ':config' => qw(bundling pass_through no_ignore_case);
use Scalar::Util 'openhandle';

use App::ForkProve::SourceHandler;

our @Blacklists = qw( Test::SharedFork );
our %Data;

sub run {
    my($class, @args) = @_;

    # Probably have to copy to @ARGV so that App::Prove can mangle it
    # in theory we don't need to, but doing so will make some tests fail
    # even with local @ARGV = () in SourceHandler. Not sure why...
    local @ARGV = @args;

    @ARGV = map { /^(-M)(.+)/ ? ($1,$2) : $_ } @ARGV;

    my @modules;
    my $lib;
    my $blib;
    my @inc;
    Getopt::Long::GetOptions('M=s@', \@modules);
    Getopt::Long::GetOptionsFromArray([@ARGV],
        'l|lib',  \$lib,
        'b|blib', \$blib,
        'I=s@',   \@inc,
    );

    if ($lib) {
        unshift @inc, 'lib';
    }

    if ($blib) {
        unshift @inc, 'blib/lib', 'blib/arch';
    }


    for (@modules) {
        my($module, @import) = split /[=,]/;
        local @INC{map pkg_to_file($_), @Blacklists} = (__FILE__) x scalar(@Blacklists);
        local @INC = (@inc, @INC);

        eval "require $module" or die $@;
        $module->import(@import);
    }

    for my $loaded (keys %INC) {
        next unless $loaded =~ /\.pm$/;

        my $mod = $loaded;
        $mod =~ s{/}{::}g;
        $mod =~ s{\.pm$}{};

        my $fh = do {
            no strict 'refs';
            *{ $mod . '::DATA' }
        };

        $Data{$mod} = [!!openhandle($fh), $INC{$loaded}];
        push @{ $Data{$mod} }, tell($fh)
            if $Data{$mod}[0];
    }

    my $app = App::Prove->new;
    $app->process_args(@ARGV);
    $app->run;
}

sub pkg_to_file {
    local $_ = shift;
    s|::|/|g;
    "$_.pm";
}


1;
__END__

=encoding utf-8

=head1 NAME

App::ForkProve - forking prove

=head1 SYNOPSIS

  use App::ForkProve;
  App::ForkProve->run(@ARGV);

=head1 DESCRIPTION

App::ForkProve is a backend for L<forkprove>.

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 COPYRIGHT

Copyright 2012- Tatsuhiko Miyagawa

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<forkprove>

=cut


lib/App/ForkProve/SourceHandler.pm
package App::ForkProve::SourceHandler;
use strict;
use parent qw(TAP::Parser::SourceHandler);

use App::ForkProve::PipeIterator;
use TAP::Parser::IteratorFactory;
TAP::Parser::IteratorFactory->register_handler(__PACKAGE__);

use TAP::Parser::SourceHandler::Perl;

sub can_handle {
    my($class, $src) = @_;

    # I think there's a TAP::Parser bug thinking the first line of .t
    # file as a shebang even if it doesn't begin with '#!'
    local $src->meta->{file}{shebang} = undef
      if $src->meta->{file}{shebang} !~ /^#!/;

    my $is_perl = TAP::Parser::SourceHandler::Perl->can_handle($src);
    return 1 if $is_perl > 0.5 && !$class->ignore($src->meta->{file});
    return 0;
}

sub ignore {
    my($class, $file) = @_;
    $ENV{PERL_FORKPROVE_IGNORE} && ($file->{dir} . $file->{basename}) =~ /$ENV{PERL_FORKPROVE_IGNORE}/;
}

sub make_iterator {
    my($class, $src) = @_;

    my $path = $src->meta->{file}{dir} . $src->meta->{file}{basename};
    my @inc = map { s/^-I//; $_ } grep { /^-I/ } @{ $src->switches };

    $class->_autoflush(\*STDOUT);
    $class->_autoflush(\*STDERR);

    pipe my $reader, my $writer;
    my $pid = fork;
    if ($pid) {
        close $writer;
        return App::ForkProve::PipeIterator->new($reader, $pid);
    } else {
        close $reader;
        open STDOUT, ">&", $writer;
        open STDERR, ">&", $writer if $src->merge;
        _run($path, \@inc);
        exit;
    }
}

sub _run {
    my ($t, $inc) = @_;

    # Many tests especially Exception tests rely on test file name being
    # passed as t/foo.t without a leading path
    local $0 = $t;
    local @INC = (@$inc, @INC);

    # if FindBin is preloaded, reset it with the new $0
    if (defined &FindBin::init) {
        FindBin::init()
    }

    # open DATA from test script
    {
        close ::DATA;
        if (open my $fh, $t) {
            my $code = do { local $/; <$fh> };
            if(my($data) = $code =~ /^__(?:END|DATA)__$(.*)/ms){
                open ::DATA, '<', \$data
                  or die "Can't open string as DATA. $!";
            }
        }
    }

    # restore defaults
    Getopt::Long::ConfigDefaults();

    # reset the state of empty pattern matches, so that they have the same
    # behavior as running in a clean process.
    # see "The empty pattern //" in perlop.
    # note that this has to be dynamically scoped and can't go to other subs
    "" =~ /^/;

    # Test::Builder is loaded? Reset the $Test object to make it unaware
    # that it's a forked off proecess so that subtests won't run
    if (defined $Test::Builder::Test) {
        $Test::Builder::Test->reset;
    }

    # avoid child processes sharing the same seed value as the parent
    srand();

    # clear @ARGV it's localized in App::ForkProve
    local @ARGV = ();

    # do() can't tell if a test can't be read or a .t's last statement
    # returned undef with $! set somewhere. Fortunately in case of
    # prove, non-readable .t will fail earlier in prove itself.
    eval q{ package main; do $t; die $@ if $@; 1 } or die $@;
}

sub _autoflush {
    my ( $class, $flushed ) = @_;
    my $old_fh = select $flushed;
    $| = 1;
    select $old_fh;
}

1;


Makefile.PL

use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "forking prove",
  "AUTHOR" => "Tatsuhiko Miyagawa <miyagawa\@bulknews.net>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "forkprove",
  "EXE_FILES" => [
    "script/forkprove"
  ],
  "LICENSE" => "perl",
  "NAME" => "forkprove",
  "PREREQ_PM" => {
    "App::Prove" => "3.25",
    "TAP::Harness" => "3.25",
    "parent" => "0.221",
    "version" => "0.77"
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.94",
    "Test::Requires" => 0
  },
  "VERSION" => "v0.4.9",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



